### DISCUSSION TITLE
4.38 configuration: unable to set the OICD jwks pivate key with a file secret

### DISCUSSION THREAD

**Discussion by thpiron** [authelia/authelia#6956] [Q&A] [answered]
Hello,

After upgrading to 4.38, I'm trying to replace all the deprecated configurations to the new ones.
But I'm stuck with this one:
```
Configuration: configuration key 'identity_providers.oidc.issuer_private_key' is deprecated in 4.38.0 and has been replaced by 'identity_providers.oidc.jwks'
```

In my current configuration I'm using  the AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE env variable to be able to use docker secret as it's recommended on the documentation.

But I can't find a way to use a file secret to store the identity_providers.oidc.jwks private key anymore.

Is there a way to do so?

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Yes, there is a way to use a file secret for the `identity_providers.oidc.jwks` private key in Authelia version 4.38.0, but it requires a specific approach due to the structure of `jwks` being an array.

Firstly, it's important to note that the environment variable `AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE` is deprecated and replaced by the use of JWKS (JSON Web Key Sets) as indicated by the deprecation warning you encountered. This change is part of the move towards a more standardized and secure method of handling keys.

To migrate your configuration to use JWKS, you should:

1. Generate your JWKS if you haven't already. This includes both the public and private keys. The private key is used by Authelia, and the public key is made available for clients to verify JWTs (JSON Web Tokens).

2. Store your JWKS in a file. This file will contain the JWKS structure, including the private key.

3. Use the environment variable `AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_FILE` to specify the path to the file containing your JWKS. This environment variable should point to the file where your JWKS is stored, allowing Authelia to load it on startup.

However, there was a reported issue ([#6861](https://github.com/authelia/authelia/issues/6861)) indicating that the `AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_FILE` environment variable was not being recognized correctly in version 4.38.0. This issue was acknowledged, and a user found that the correct environment variable to use was `AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_KEY_FILE`, but this was a misunderstanding as clarified later in the discussion. The correct approach is to use the `AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_FILE` environment variable, but ensure your JWKS file is correctly formatted and located in the path specified by the environment variable.

If you encounter issues with this approach, it might be due to the reported bug or a misconfiguration. Ensure your JWKS file is correctly formatted and accessible by Authelia. If the problem persists, consider reaching out to Authelia's support channels or checking for updates that might address this issue.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment, marked as answer]
In the release notes under the template option: https://www.authelia.com/blog/4.38-release-notes/#multiple-json-web-keys

You just need to enable the `template` filter.

**Reply by thpiron** [reply]
My bad, I missed it when reading the release notes.
Thank you for your quick answer!

**Comment by auspex** [comment]
But…

The documentation states that I can use a `secrets` file for any variable ending in `key`. And since I'm already providing secrets for four other Authelia config variables that way, I was rather resistant to using templates. Nowhere on the page https://www.authelia.com/configuration/identity-providers/openid-connect/provider does it say that _"the new configuration style does not natively support secrets"_ (which is what the release notes say).

**Reply by james-d-elliott** [reply]
Clarified (we don't and wont support it): https://www.authelia.com/configuration/methods/secrets/#environment-variables
